Add extended Euclidean algorithm

Multramate opened the pull request:
Added an extended Euclidean algorithm function `extEuclid` in `Data/Poly/Internal/Dense.hs` for `Num` and `Semiring` polynomials to partially address https://github.com/Bodigrim/poly/issues/22, with several remarks:
- I do not know if it is possible to have a general extended GCD algorithm for all GCD domains, so `Euclidean` is added as a constraint here for the use of `quot`.
- This module is probably not the best place to add GCD-related functions, but since `Euclidean` in `semirings` does not have this as part of its class, I figured that putting it here would suffice for now.
- The current implementation is completely unoptimised, so it would be great if you could shed some light on the actual efficiency of the function.
- There is currently no equivalent implementation for `PolyOverFractional` polynomials.
- There is currently no equivalent implementation for sparse polynomials.
NB: Wouldn't it be nice if there is a `Ring k => Field k` class? This way all polynomials over fields are automatically Euclidean domains, allowing for a `semirings` version of `PolyOverFractional`.

Added NFData instances derived from `Vector` as well.

Bodigrim commented:
That's nice! Sorry, I am kinda living on the road these days.

1. Could you possibly move `NFData` changes to a separate PR? I would merge them instantly, while the extended gcd requires some pondering about.

2. Please name the main routine `gcdExt`, in spirit of `GHC.Integer.GMP.Internals.gcdExtInteger`.

3. I would like to see more detailed haddock comments and tests written. Simple benchmarks, along the lines of `DenseBench.benchGcd` are also welcome.

4. You are right, in general it is not possible to have `gcdExt` for a `GcdDomain`, which is not `Euclidean`. For example, `gcd (X :: UPoly Int) (2 :: UPoly Int) = 1`, but there are no `a, b :: Int` such that `a * X + b * 2 = 1`. So imposing `Euclidean` constraint is totally justified.

5. There is a [discussion](https://github.com/chessai/semirings/issues/36) about `Field` type class in `semirings`. May I ask you about the reasons to use `Field` instead of plain old `Fractional` in `galois-field`?

Multramate commented:
Separated NFData PR.

Multramate commented:
I have renamed the routine as `gcdExt` as per suggested, and have also modified its type to only return `g` and `s` in `as + bt = g` rather than `g`, `s`, and `t` to mimic `gcdExtInteger`.

To allow for a canonical form for `gcdExt`, I have also added the functions `fractionalGcdExt` (in an attempt to mirror `gcdExt` and `fractionalGcd` in terms of naming) and `scaleMonic` in `Data/Poly/Internal/Dense/Fractional.hs`. These are currently exposed functions used for testing and convenience, but could be deleted later if deemed unnecessary.

More detailed Haddock comments are now available for all of these functions, as well as simple tests and benchmarks mimicking those already available. Let me know if they are in any way incomplete or if I may have wrote them incorrectly somewhere (and I think I did so for benchmarks, since `gcdExt` seems to be outperforming `gcd` and `gcdFrac` by a large margin).

My main reason for using `Field` rather than `Fractional` is that it makes more mathematical sense (and hence a personal preference) in abstracting a common `Field` interface that could be used for non-finite fields. Experimentally, I also found the `Semiring` interface for `Poly` to be slightly faster, and so decided to switch everything over, but still expose a `Num` interface for compatibility. Thank you for the link to the discussion in the `semirings` library.

(On a side note, I found out that extended GCDs are well-defined for so-called Bezout domains, which include all principal ideal domains (and hence all Euclidean domains) but not all unique factorisation domains (and hence not all GCD domains). Do you have thoughts as to why the `GcdDomain` and `Euclidean` classes are not defined with `Ring` as a constraint? This would make several things simpler here.)

Bodigrim commented:
> Do you have thoughts as to why the `GcdDomain` and `Euclidean` classes are not defined with `Ring` as a constraint?

Because `Natural` has reasonable `GcdDomain` and `Euclidean` instances, but is not a `Ring`.

---

There are certainly some compromises in the design of `GcdDomain`/`Euclidean`.

The thing is that even if we try our best, __we cannot faithfully recreate mathematical hierarchy.__ Any property of existence (like, for any `x` and `y` there exists `gcd x y`) cannot be rendered into Haskell otherwise than a constructive proof (like, here is an algorithm to compute `gcd x y`). Such algorithms are not readily available, because mathematical proofs of existence are free to be non-constructive.

A notable example is `Data.Euclidean.GcdDomain.divide`. What we actually want to express is the notion of an [integral domain](https://en.wikipedia.org/wiki/Integral_domain): for each non-zero `a` and `b` there exists no more than one solution to `a * x == b`. Mathematically this property is easy to check, because it is equivalent to the statement that the product of any two nonzero elements is nonzero. But the only way to say this in Haskell is to provide a division algorithm `a -> a -> Maybe a`, which is a way more difficult accomplishment.

That said, my _personal_ preference to the numerical hierarchy in Haskell is that if we cannot achieve mathematical purity, we should be driven mostly by practical considerations and usability. 

---

I'll get back to the review on Monday, sorry for the delay.

Multramate commented:
I see - `Natural` not being a `Ring` makes sense. I completely agree with your last stance.

Although I suppose the class names could have been chosen differently to emphasise this - e.g. `GCDSemiring` and `EuclideanSemiring` - since (integral) domains are by definition commutative rings. The notion of having `divide` could also be a class `DivisibleSemiring` or `IntegralSemiring` by itself, without requiring GCD operations.

Bodigrim commented:
I think `EuclideanSemiring` sounds too scary for an average software engineer. 

There was a temptation to introduce `IntegralDomain` as a superclass of `GCDDomain`, but I struggled to find an example of Haskell data type, which is `IntegralDomain`, but not `GCDDomain`.

Bodigrim commented:
Back to review. There are now three functions with slightly different constraints:

```
gcdExt 
  :: (Eq (Poly v a), E.Euclidean (Poly v a), Num (Poly v a))
  => Poly v a -> Poly v a -> (Poly v a, Poly v a)

gcdExt' 
  :: (Eq (Poly v a), E.Euclidean (Poly v a), Semiring.Ring (Poly v a))
  => Poly v a -> Poly v a -> (Poly v a, Poly v a)

fractionalGcdExt
  :: (Eq a, Fractional a, GcdDomain a, Semiring.Ring a, G.Vector v a, Eq (v a))
  => Poly v a -> Poly v a -> (Poly v a, Poly v a)
```

Such diversity is confusing. Note that the only way to have `E.Euclidean (Poly v a)` is to be `Fractional`:

```haskell
instance (Eq a, Eq (v a), Ring a, GcdDomain a, Fractional a, Vector v a) => Euclidean (Poly v a)
```

It means that if `Poly v a` satisfies constraints of `gcdExt` or `gcdExt'`, it also neccessarily satisfies constraints for `fractionalGcdExt`. So let us merge them together and expose only one function from `Data.Poly.Internal.Dense`:

```haskell
gcdExt
  :: (Eq a, Fractional a, GcdDomain a, Semiring.Ring a, G.Vector v a, Eq (v a))
  => Poly v a -> Poly v a -> (Poly v a, Poly v a)
```

It also eliminates reasons to have `Data.Poly.Internal.PolyOverFractional.fractionalGcdExt`: end users can easily `Data.Coerce.coerce gcdExt`.

--- 

Great tests and benchmarks!

Multramate commented:
> I think `EuclideanSemiring` sounds too scary for an average software engineer.

Well this is certainly subjective, and anyone who intends to use this class should know what it is and what it is not. Having an accurate mathematical definition (just `Euclidean` is kind of ambiguous, and `GcdDomain` is plain wrong) will at least make this fully clear - it wasn't clear to me initially, and a similar query may arise in the future by others intending to expand the library.

> There was a temptation to introduce `IntegralDomain` as a superclass of `GCDDomain`, but I struggled to find an example of Haskell data type, which is `IntegralDomain`, but not `GCDDomain`.

That makes sense, but other than non-fractional polynomials, are there any other Haskell examples of GCD semirings that are not Euclidean semirings?

Multramate commented:
> Note that the only way to have E.Euclidean (Poly v a) is to be Fractional:

Yeah, my bad, thanks - will merge those three together.

Bodigrim commented:
> That makes sense, but other than non-fractional polynomials, are there any other Haskell examples of GCD semirings that are not Euclidean semirings?

Currently I am not aware of such examples :) `Data.Poly` evolved as a spin-off of `semirings` and was carefully co-designed together with `Data.Euclidean`. 

---

Merged, thanks for your efforts!
